### ISSUE TITLE
Encrypting with constants.RSA_NO_PADDING yields "data too small for key size"

### ISSUE THREAD

**Issue by Ottunger** [rzcoder/node-rsa#72]
Hi,
As I want to interoperate with another library that uses weird padding, I've been disabling classic padding to do mine beforehand. But the encryption yields this error although the key is 1024 bits, and the data I try to encrypt is 128bytes = 1024bits long as well.

    var dec = new RSA(
        rsa_key, 'pkcs8-public', {
            encryptionScheme: {
                scheme: 'pkcs1',
                padding: constants.RSA_NO_PADDING
            }
        }
    );
    var arr = pkcs1pad2(data, 1024);
    console.log(arr, dec.getMaxMessageSize(), arr.length);
    try {console.log(dec.encrypt(arr));}catch(e){console.log(e);}

Yields: my padded message, 128, 128, error:0406B07A:rsa routines:RSA_padding_add_none:data too small for key size.

Worst is: if I try to encrypt the plaintext non padded, and let your code pad it with zeros from what I see, I get the same error!

Is there something I haven't understood?

**Comment by rzcoder** [comment]
Can you provide full code with your key & data?

**Comment by Ottunger** [comment]
Of course. For the sake of interest here is first pkcs1pad2, but as I said using or not does not change anything.

/**
     * Pads for PKCS1type2
     * @function pkcs1pad2
     * @public
     * @param {Number[]} b Data.
     * @param {Number} k Number of bits in key.
     * @return {Number[]} Padded.
     */
    export function pkcs1pad2(s: number[], k: number): number[] {
        var n = (k + 7) >> 3;
        if(n < s.length + 11) {
            return null;
        }
        var ba = new Array(n);
        var i = s.length - 1;
        while(i >= 0 && n > 0) {
            var c = s[i--];
            if(c < 128) {
                ba[--n] = c;
            } else {
                ba[--n] = (c & 63) | 128;
                ba[--n] = (c >> 6) | 192;
            }
        }
        ba[--n] = 0;
        while(n > 2) {
            ba[--n] = 0;
            while(ba[n] == 0)
                ba[n] = Math.floor(Math.random() * 254 + 1);
        }
        ba[--n] = 2;
        ba[--n] = 0;
        return ba;
    }

    console.log(data, rsa_key);
    var dec = new RSA(
        rsa_key, 'pkcs8-public', {
            encryptionScheme: {
                scheme: 'pkcs1',
                padding: constants.RSA_NO_PADDING
            }
        }
    );
    var arr = pkcs1pad2(data, 1024);
    console.log(arr);
    try {
        console.log(dec.encrypt(arr));
    } catch(e) {
        console.log(e);
    }

Yields:
[17, 105, 21, 237, 144, 244, 90, 203, 201, 47, 48, 59, 85, 218, 217, 223, 228, 187, 222, 198, 38, 222, 239, 62, 80, 209, 244, 230, 211, 111, 65, 16]

'-----BEGIN PUBLIC KEY-----\nMIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCsV3U9BKwt3Wgy/N2HpI/qaVXV\n9zYjWXyuO/IpcEa0FIPH8LOSca0pk9ab/8fufAoohjTPPfcB0bmLf9X8wP9ZRHBG\ntUICIjvHNywPfoQ9/E/m6ebP/Ke6yDqBaPBx/pTiaUmBOSS0/EfDIiwNGiJz6xua\nBwn0S3SdIzFIZVjXiwIDAQAB\n-----END PUBLIC KEY-----'

[ 0, 2, 25, 117, 147, 240, 173, 20, 22, 149, 221, 111, 206, 123, 106, 94, 127, 18, 127, 47, 225, 214, 182, 128, 110, 105, 186, 165, 15, 228, 47, 172, 238, 174, 140, 65, 203, 40, 31, 167, 141, 118, 160, 232, 30, 238, 183, 179, 153, 134, 96, 194, 64, 149, 177, 120, 97, 249, 73, 18, 96, 69, 34, 233, 130, 89, 49, 52, 218, 35, 219, 19, 180, 166, 35, 74, 211, 0, 17, 105, 21, 195, 173, 194, 144, 195, 180, 90, 195, 139, 195, 137, 47, 48, 59, 85, 195, 154, 195, 153, 195, 159, 195, 164, 194, 187, 195, 158, 195, 134, 38, 195, 158, 195, 175, 62, 80, 195, 145, 195, 180, 195, 166, 195, 147, 111, 65, 16 ]

[Error: Error during encryption. Original error: Error: error:0406B07A:rsa routines:RSA_padding_add_none:data too small for key size]

If you need more info, I'm available, thanks!

**Comment by rzcoder** [comment]
Why you use pkcs1 padding with RSA_NO_PADDING option? If you need `pkcs1` just pass 
```
encryptionScheme: {
   scheme: 'pkcs1'
}
```

**Comment by rzcoder** [comment]
Also, when you try encode an js Array or Object, lib will cast it to JSON string. Use `dec.encrypt(Buffer.from(arr))`

**Comment by Ottunger** [comment]
Because I must interoperate it with a service that uses https://github.com/travist/jsencrypt, and it only uses as padding pkcs1type2, the function that I copied. I do not know it this is the standard pkcs1 padding, but anyways, it does not interoperate with any padding I have found so far, would it be a PHP lib or whatever, yours included. So I first pad the data using jsencrypt's padding, and only then encrypt.

But you are right, if I let your library pad the already padded data, this works fine. 

To sum up:
    var dec = new RSA(
        rsa_key, 'pkcs8-public', {
            encryptionScheme: {
                scheme: 'pkcs1'
            }
        }
    );
    var arr = pkcs1pad2(data, 1024);
    return dec.encrypt(arr);

works for me. But maybe it would be interesting to make sure that RSA_NO_PADDING always works as  expected, as I have not been able to make it work while encrypting. Decryption works fine, though.

Thanks for the remark as well.
